### PR TITLE
Show unread count in room's back button

### LINKED ISSUE #778: [NEW] Show unread count in room's back button

### PR DESCRIPTION
**Pull request by djorkaeffalexandre**
@RocketChat/ReactNative
@diegolmello 

Closes #778 

Printscreen:

### CODE REVIEW

File: app/views/RoomView/index.js [hunk 1]
```diff
@@ -48,6 +48,7 @@ import { COLOR_WHITE } from '../../constants/colors';
 	actionsShow: actionMessage => dispatch(actionsShowAction(actionMessage)),
 	closeRoom: () => dispatch(closeRoomAction())
 }))
+
```
Comment (diegolmello) [inline app/views/RoomView/index.js]: Unnecessary empty space

File: app/views/RoomsListView/index.js [hunk 2]
```diff
@@ -67,8 +67,9 @@ export default class RoomsListView extends LoggedView {
 	static navigationOptions = ({ navigation }) => {
 		const searching = navigation.getParam('searching');
 		const cancelSearchingAndroid = navigation.getParam('cancelSearchingAndroid');
-		const onPressItem = navigation.getParam('onPressItem', () => {});
-		const initSearchingAndroid = navigation.getParam('initSearchingAndroid', () => {});
+		const onPressItem = navigation.getParam('onPressItem', () => { });
```
Comment (diegolmello) [inline app/views/RoomsListView/index.js]: Unnecessary empty space

File: app/views/RoomsListView/index.js [hunk 3]
```diff
@@ -67,8 +67,9 @@ export default class RoomsListView extends LoggedView {
...
+		const initSearchingAndroid = navigation.getParam('initSearchingAndroid', () => { });
```
Comment (diegolmello) [inline app/views/RoomsListView/index.js]: Unnecessary empty space

File: app/views/RoomsListView/index.js [hunk 4]
```diff
@@ -140,7 +142,9 @@ export default class RoomsListView extends LoggedView {
 		this.getSubscriptions();
 		const { navigation } = this.props;
 		navigation.setParams({
-			onPressItem: this._onPressItem, initSearchingAndroid: this.initSearchingAndroid, cancelSearchingAndroid: this.cancelSearchingAndroid
+			onPressItem: this._onPressItem,
```
Comment (diegolmello) [inline app/views/RoomsListView/index.js]: Unnecessary update

File: app/views/RoomsListView/index.js [hunk 5]
```diff
@@ -580,7 +595,7 @@ export default class RoomsListView extends LoggedView {
 
 		const { showUnread, showFavorites, groupByType } = this.props;
 		if (!(showUnread || showFavorites || groupByType)) {
-			const { chats, search } = this.state;
+			const { search, chats } = this.state;
```
Comment (diegolmello) [inline app/views/RoomsListView/index.js]: Unnecessary update

File: app/views/RoomsListView/index.js [hunk 6]
```diff
@@ -328,6 +332,17 @@ export default class RoomsListView extends LoggedView {
 				this.removeListener(this.livechat);
 			}
 
+			// set Unread Count
+			this.data.addListener(debounce(() => {
```
Comment (diegolmello) [inline app/views/RoomsListView/index.js]: You should search for unread count in RoomView's componentDidMount.

Comment (djorkaeffalexandre) [reply app/views/RoomsListView/index.js]: You want only unread count on componentDidMount?

File: app/views/RoomsListView/index.js (hunk 4 above)
Comment (djorkaeffalexandre) [reply app/views/RoomsListView/index.js]: With more one param this need object-curly-newline

File: app/views/RoomsListView/index.js (hunk 6 above)
Comment (djorkaeffalexandre) [reply app/views/RoomsListView/index.js]: The headerBackTitle has to be defined in the origin screen, not in the destination screen. If I search for unread count in RoomView I don't can set that.

Comment (djorkaeffalexandre) [reply app/views/RoomsListView/index.js]: I moved the listener to ComponentDidMount, is that?

Comment (diegolmello) [reply app/views/RoomsListView/index.js]: If you declare `renderLeft` on RoomViewHeader, you can pass the button title you want.
Querying on RoomsListView won't work on deep link/notifications tap and won't update automatically.

Comment (djorkaeffalexandre) [reply app/views/RoomsListView/index.js]: Hello @diegolmello , in the version we are using of react-navigation the title prop of the HeaderBackButton isn't works, I have opened an issue in the react-navigation repository and I'm waiting for replies. Regards, Djorkaeff.

**Comment by djorkaeffalexandre** [comment]
@diegolmello I solved your reviews.

File: app/views/RoomView/index.js [hunk 7]
```diff
@@ -131,6 +141,14 @@ export default class RoomView extends LoggedView {
 				}
 			}
 			safeAddListener(this.rooms, this.updateRoom);
+			safeAddListener(this.chats, () => {
+				const unreadsCount = this.chats.filtered('(unread > 0 || alert == true)').length;
+				if (unreadsCount !== navigation.getParam('unreadsCount')) {
+					navigation.setParams({
+						unreadsCount
```
Comment (diegolmello) [inline app/views/RoomView/index.js]: Unread count should consider the sum of `unread` and `userMentions` of all subscriptions.

Comment (diegolmello) [reply app/views/RoomView/index.js]: You can compare against web badge.
Example:

File: app/views/RoomView/index.js [hunk 8]
```diff
@@ -131,6 +141,14 @@ export default class RoomView extends LoggedView {
 				}
 			}
 			safeAddListener(this.rooms, this.updateRoom);
+			safeAddListener(this.chats, () => {
```
Comment (diegolmello) [inline app/views/RoomView/index.js]: Create a function with debounce of 300ms.

File: app/views/RoomView/index.js (hunk 1 above)
Comment (diegolmello) [reply app/views/RoomView/index.js]: Delete this empty line.

**Comment by diegolmello** [comment]
@djorkaeffalexandre Also resolve conflicts and merge develop. There's a few changes on RoomView header to make it prettier.
The title is aligned left now and you'll need to make a few layout ajustments.

**Comment by djorkaeffalexandre** [comment]
Gif with header back title:

**Comment by djorkaeffalexandre** [comment]
Printscreen to unreadCount on Android Device:

**Comment by diegolmello** [comment]
> It works perfectly for me, I just want to point out that in web If some channel is unread and I am not mentioned in it, it shows (.) type of indicator can you also add that

No need for the dot behaviour, but thanks for reviewing :)

File: app/views/RoomView/index.js [hunk 9]
```diff
@@ -67,18 +67,28 @@ export default class RoomView extends LoggedView {
...
-				<RoomHeaderView
-					rid={rid}
-					prid={prid}
-					tmid={tmid}
-					title={title}
-					type={t}
-					widthOffset={tmid ? 95 : 130}
-					isFetching={isFetching}
-				/>
+			headerLeft: (
```
Comment (diegolmello) [inline app/views/RoomView/index.js]: Placing the entire header on the left button feels wrong and may be impacted by future react-navigation updates.

File: app/views/RoomView/index.js [hunk 10]
```diff
@@ -67,18 +67,28 @@ export default class RoomView extends LoggedView {
...
+				<View style={styles.headerLeftContainerStyle}>
+					<View style={styles.headerBackButton}>
+						<HeaderBackButton
+							title={unreadsCount > 999 ? '+999' : unreadsCount}
+							backTitleVisible
+							onPress={() => navigation.goBack()}
```
Comment (diegolmello) [inline app/views/RoomView/index.js]: Can we use `onPress={navigation.goBack}`?

**Pull request closed without merging**